keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,299 @@
+
+Section below describes behaviour of old predicates, CSI predicate and scheduler after the proposal has been implemented.
+For brevity - "old predicates" refers to now deprecated cloudprovider specific predicates - `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount`.
+
+#### Implementation Detail for all CSI Drivers
+
+* Kubelet will create `CSINode` instance during initial CSI Driver registration.
+  * Limits of each CSI volume plugin will be added to `CSINode.spec.drivers[xyz].allocatable`.
+  * User may NOT change `CSINode.spec.drivers[xyz].allocatable` to override volume plugin / CSI driver values, e.g. to "reserve" some attachment to the operating system. Kubelet will periodically reconcile `CSINode` and overwrite the value.
+    * Especially, `kubelet --kube-reserved` or `--system-reserved` cannot be used to "reserve" volumes for kubelet or the OS. It is not possible with existing kubelet and this KEP does not change it. We expect that CSI drivers will have configuration options / cmdline arguments to reserve some volumes and they will report their limit already reduced by that reserved amount.
+* Scheduler will respect `Node.status.allocatable` and `Node.status.capacity` for CSI volumes if `CSINode` object is not available or has missing entry in `CSINode.spec.drivers[xyz].allocatable` during a deprecation period but kubelet will stop populating `Node.status.allocatable` and `Node.status.capacity` for CSI volumes.
+  * After deprecation period for CSI volumes, limits coming from `Node.status.allocatable` and `Node.status.capacity` will be completely ignored.
+  * After deprecation period, scheduler won't schedule any pods that use CSI volumes to a node with missing `CSINode` instance. It is expected that it happens only during node registration when `Node` exists and `CSINode` doesn't and it self-heals quickly.
+  * After deprecation period, if `CSINode` exists and  does not contain a CSI driver, scheduler will not put CSI volumes for the driver to the node.
+
+
+#### Implementation detail for in-tree Drivers with CSI migration disabled
+
+##### When no CSI driver for same underlying storage type is installed on the node.
+
+* For Azure, GCEPD, AWS and Cinder - in-tree volume plugins will keep reporting their limits via `Node` object and old predicates will work as expected until CSI migration has been enabled (and GA) for given volume plugin.
+
+##### When CSI driver for same underlying storage type is installed on the node.
+
+* For Azure, GCEPD, AWS and Cinder - in-tree volume plugins will report their limits via `Node` object same as before.
+* Old predicates will be modified to perform an additional check of `CSINode` at very beginning of their code. If they detect that `CSINode` object contains limit of same underlying volume type, they will return early(with success) and `MaxCSIVolumeCountPred` will be responsible for counting both CSI and in-tree volumes of same type.
```

Comment:
Clarify here that it needs to use the migration library to translate the in-tree name to the csi name.  And this is done even when migration is off.

Reply:
There is a section below that covers it:

> [CSI migration library](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/csi-translation-lib) is used to find CSI driver name for in-tree volume plugins + its `VolumeHandle`. This CSI driver name is used as key in `CSINode.CSINode.spec.drivers[xyz].allocatable` list. The `VolumeHandle` is unique for each volume and will be used to de-duplicate volumes used by multiple pods on the same node.